core/vm/contracts.go

```diff
@@ -88,47 +101,234 @@ func (c *ecrecover) Run(in []byte) ([]byte, error) {
 type sha256hash struct{}
 
 // RequiredGas returns the gas required to execute the pre-compiled contract.
 //
 // This method does not require any overflow checking as the input size gas costs
 // required for anything significant is so high it's impossible to pay for.
 func (c *sha256hash) RequiredGas(input []byte) uint64 {
-	return uint64(len(input)+31)/32*params.Sha256WordGas + params.Sha256Gas
+	return uint64(len(input)+31)/32*params.Sha256PerWordGas + params.Sha256BaseGas
 }
-func (c *sha256hash) Run(in []byte) ([]byte, error) {
-	h := sha256.Sum256(in)
+func (c *sha256hash) Run(input []byte) ([]byte, error) {
+	h := sha256.Sum256(input)
 	return h[:], nil
 }
 
-// RIPMED160 implemented as a native contract
+// RIPMED160 implemented as a native contract.
 type ripemd160hash struct{}
 
 // RequiredGas returns the gas required to execute the pre-compiled contract.
 //
 // This method does not require any overflow checking as the input size gas costs
 // required for anything significant is so high it's impossible to pay for.
 func (c *ripemd160hash) RequiredGas(input []byte) uint64 {
-	return uint64(len(input)+31)/32*params.Ripemd160WordGas + params.Ripemd160Gas
+	return uint64(len(input)+31)/32*params.Ripemd160PerWordGas + params.Ripemd160BaseGas
 }
-func (c *ripemd160hash) Run(in []byte) ([]byte, error) {
+func (c *ripemd160hash) Run(input []byte) ([]byte, error) {
 	ripemd := ripemd160.New()
-	ripemd.Write(in)
+	ripemd.Write(input)
 	return common.LeftPadBytes(ripemd.Sum(nil), 32), nil
 }
 
-// data copy implemented as a native contract
+// data copy implemented as a native contract.
 type dataCopy struct{}
 
 // RequiredGas returns the gas required to execute the pre-compiled contract.
 //
 // This method does not require any overflow checking as the input size gas costs
 // required for anything significant is so high it's impossible to pay for.
 func (c *dataCopy) RequiredGas(input []byte) uint64 {
-	return uint64(len(input)+31)/32*params.IdentityWordGas + params.IdentityGas
+	return uint64(len(input)+31)/32*params.IdentityPerWordGas + params.IdentityBaseGas
 }
 func (c *dataCopy) Run(in []byte) ([]byte, error) {
 	return in, nil
 }
+
+// bigModExp implements a native big integer exponential modular operation.
+type bigModExp struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+func (c *bigModExp) RequiredGas(input []byte) uint64 {
+	// Pad the input with zeroes to the minimum size to read the field lengths
+	input = common.RightPadBytes(input, 96)
+
+	var (
+		baseLen = new(big.Int).SetBytes(input[:32])
+		expLen  = new(big.Int).SetBytes(input[32:64])
+		modLen  = new(big.Int).SetBytes(input[64:96])
+	)
+	input = input[96:]
+
+	// Retrieve the head 32 bytes of exp for the adjusted exponent length
+	var expHead *big.Int
+	if big.NewInt(int64(len(input))).Cmp(baseLen) <= 0 {
+		expHead = new(big.Int)
+	} else {
+		offset := int(baseLen.Uint64())
+
+		input = common.RightPadBytes(input, offset+32)
```

Comment:
This will copy the entire input. Is that really necessary? 
Why not just use `getData`, as in either
```
expHead := new(big.Int).SetBytes(getData(input, baseLen, expLen))
```
or 
```
expHead := new(big.Int).SetBytes(getData(input, baseLen, common.Big32))
```